test/apiserver/module_tests/test_count_importance.py

```diff
@@ -16,8 +17,8 @@
 
 def main():
     cases = [
-        (POSITIVE_MARKERS, NEGATIVE_MARKERS, 3),
-        (NEGATIVE_MARKERS, POSITIVE_MARKERS, 168),
+        (POSITIVE_MARKERS, NEGATIVE_MARKERS, 7),
+        (NEGATIVE_MARKERS, POSITIVE_MARKERS, 352),
```

Comment:
We're only retrieving the top 50 most important cells from each specimen, who each have 100 cells. The fake importance scores I created were distributed randomly between 0 and 1, and I manually verified the exact count that we should see given them by merging the importance CSV with `0.csv` to `6.csv`. 3 and 168 should be correct.

Reply:
Gotcha, I wasn't sure if you just added up what came back.
Yeah what is coming back now isn't right, as there are many samples whose counts are greater than 50. So the cell limit isn't working. Working on it now.

Reply:
I'm getting more reasonable numbers now for that test, but they don't match what you say you got by manual inspection. Can you post a short snippet of deriving the expected values here?

Reply:
Here is my snippet for generating the expected counts:
```py
from json import dumps

from pandas import read_csv
from pandas import Series

def get_file_by_sample():
    filename = 'test/test_data/adi_preprocessed_tables/dataset1/file_manifest.tsv'
    manifest = read_csv(filename, sep='\t')
    rows = [row for _, row in manifest.iterrows() if row['Data type'] == 'Tabular cell manifest']
    return {
        row['Sample ID']: row['File name']
        for row in rows
    }

def get_counts(signature, expression_dfs, importances):
    offset = 0
    counts = {}
    for filename, df in expression_dfs.items():
        df['histological_structure'] = [str(offset + i) for i in range(100)]
        df.set_index('histological_structure', inplace=True)
        offset = offset + 1
        joined = df.join(importances, on='histological_structure')
        joined = joined.sort_values(by='importance_score', ascending=True)
        rank = Series([i+1 for i in range(joined.shape[0])], joined.index)
        df['rank'] = rank
        cell_limit = 50
        df2 = df.loc[df['rank'] <= cell_limit]
        assert df2.shape[0] == cell_limit
        conditions = [df2[channel_name] == value for channel_name, value in signature]
        condition = conditions[0] & conditions[1] & conditions[2]
        selection = df2[condition]
        counts[filename] = selection.shape[0]
    counts_by_sample = {
        sample: counts[get_file_by_sample()[sample]]
        for sample in get_file_by_sample()
    }
    return counts_by_sample

def generate():
    importances_filename = 'test/test_data/adi_preprocessed_tables/dataset1/importances.csv'
    importances = read_csv(importances_filename, sep=',')
    importances['histological_structure'] = importances['histological_structure'].astype(str)
    importances.set_index('histological_structure', inplace=True)
    expression_dfs = {
        f'{i}.csv': read_csv(
            f'test/test_data/adi_preprocessed_tables/dataset1/{i}.csv',
            sep=',',
            keep_default_na=False,
        )
        for i in range(7)
    }

    signature = [('CD3_Positive', 1), ('CD4_Positive', 1), ('CD8_Positive', 1)]
    counts_by_sample = get_counts(signature, expression_dfs, importances)
    print(dumps(counts_by_sample, indent=4))
    print(sum(count for _, count in counts_by_sample.items()))

    print('')

    signature = [('CD3_Positive', 0), ('CD4_Positive', 0), ('CD8_Positive', 0)]
    counts_by_sample = get_counts(signature, expression_dfs, importances)
    print(dumps(counts_by_sample, indent=4))
    print(sum(count for _, count in counts_by_sample.items()))

generate()
```

```txt
{
    "lesion 0_1": 1,
    "lesion 0_2": 2,
    "lesion 0_3": 0,
    "lesion 6_1": 0,
    "lesion 6_2": 0,
    "lesion 6_3": 0,
    "lesion 6_4": 0
}
3

{
    "lesion 0_1": 30,
    "lesion 0_2": 21,
    "lesion 0_3": 24,
    "lesion 6_1": 23,
    "lesion 6_2": 30,
    "lesion 6_3": 19,
    "lesion 6_4": 25
}
172
```

I'm getting a slightly different total (172 for the second case, not 168). Can you check against your manual derivation?

Also can you confirm that the lower values for the importance scores are "more important", i.e. cell limit of 1 should retrieve the cell with the lowest importance score?

Reply:
I might be assigning identifiers in an undefined way due to that dict enumeration. Hold on.